### PR TITLE
Implement query and get/put for example

### PR DESCRIPTION
**Pull request by Hoverbear**
This is in progress but I'll keep pushing up to this branch.

Work remaining:
* [x] Handling queries from a client on the server.
* [x] Determining if it's appropriate for `query` and `proposal` should have the same response variant. For discussion with @danburkert.

### CODE REVIEW

**Comment by Hoverbear** [comment]
@danburkert Should be ready for an r+ if you can review it sometime. :)

**Comment by Hoverbear** [comment]
I included a proof of bug for #68 in 4d42c30

**Comment by Hoverbear** [comment]
Ok, I think I'm finally done here. :)

File: examples/register.rs [hunk 1]
```diff
@@ -105,12 +106,30 @@ fn server(args: &Args) {
     Server::run(id, addr, peers, store, state_machine).unwrap();
 }
 
-fn get(_args: &Args) {
-    panic!("unimplemented: waiting on changes to the Raft Client and StateMachine APIs");
+fn get(args: &Args) {
+    let cluster = args.arg_server_address.iter()
+        .map(|v| {
```
Comment (danburkert) [inline examples/register.rs]: I think you can reduce this to `.map(parse_addr)`.

File: examples/register.rs [hunk 2]
```diff
@@ -105,12 +106,30 @@ fn server(args: &Args) {
...
+    let mut client = Client::new(cluster);
+    let response = client.query(args.arg_new_value.as_bytes()).unwrap();
+    println!("{:?}", String::from_utf8(response))
 }
 
-fn put(_args: &Args) {
-    panic!("unimplemented: waiting on changes to the Raft Client and StateMachine APIs");
+fn put(args: &Args) {
+    let cluster = args.arg_server_address.iter()
+        .map(|v| {
```
Comment (danburkert) [inline examples/register.rs]: same here.

File: examples/register.rs [hunk 3]
```diff
@@ -105,12 +106,30 @@ fn server(args: &Args) {
...
+            // println!("{:?}", v);
+            parse_addr(&v)
+        })
+        .collect();
+    // println!("{:?}", cluster);
```
Comment (danburkert) [inline examples/register.rs]: this looks leftover

File: examples/register.rs [hunk 4]
```diff
@@ -105,12 +106,30 @@ fn server(args: &Args) {
...
+    let mut client = Client::new(cluster);
+    let response = client.propose(args.arg_new_value.as_bytes()).unwrap();
+    println!("{:?}", String::from_utf8(response))
```
Comment (danburkert) [inline examples/register.rs]: I think this should be `println!("{}", String::from_utf8(response).unwrap())`

File: examples/register.rs [hunk 5]
```diff
@@ -105,12 +106,30 @@ fn server(args: &Args) {
...
+    let cluster = args.arg_server_address.iter()
+        .map(|v| {
+            // println!("{:?}", v);
+            parse_addr(&v)
+        })
+        .collect();
+    // println!("{:?}", cluster);
+    let mut client = Client::new(cluster);
+    let response = client.query(args.arg_new_value.as_bytes()).unwrap();
+    println!("{:?}", String::from_utf8(response))
```
Comment (danburkert) [inline examples/register.rs]: `println!("{}", String::from_utf8(response).unwrap())`

File: examples/register.rs [hunk 6]
```diff
@@ -105,12 +106,30 @@ fn server(args: &Args) {
...
-fn get(_args: &Args) {
-    panic!("unimplemented: waiting on changes to the Raft Client and StateMachine APIs");
+fn get(args: &Args) {
+    let cluster = args.arg_server_address.iter()
+        .map(|v| {
+            // println!("{:?}", v);
+            parse_addr(&v)
+        })
+        .collect();
+    // println!("{:?}", cluster);
```
Comment (danburkert) [inline examples/register.rs]: this looks leftover

File: src/client.rs [hunk 7]
```diff
@@ -43,10 +43,21 @@ impl Client {
...
     /// return once the entry has been durably committed.
     /// Returns `Error` when the entire cluster has an unknown leader. Try proposing again later.
-    pub fn propose(&mut self, entry: &[u8]) -> Result<()> {
+    pub fn propose(&mut self, entry: &[u8]) -> Result<Vec<u8>> {
         scoped_trace!("{:?}: propose", self);
         let mut message = messages::proposal_request(entry);
+        self.send_message(&mut message)
+    }
+
+    /// Queries an entry from the log. This is non-mutating and doesn't go through the durable log.
```
Comment (danburkert) [inline src/client.rs]: This queries the state machine, not the log directly.

File: src/replica.rs [hunk 8]
```diff
@@ -481,34 +490,69 @@ impl <S, M> Replica<S, M> where S: Store, M: StateMachine {
         scoped_debug!("proposal from Client({})", from);
 
         if self.is_candidate() || (self.is_follower() && self.follower_state.leader.is_none()) {
-            actions.client_messages.push((from.into(), messages::proposal_response_unknown_leader()));
+            scoped_debug!("proposal recieved, but leader is not known");
```
Comment (danburkert) [inline src/replica.rs]: misspelled received

File: src/replica.rs [hunk 9]
```diff
@@ -481,34 +490,69 @@ impl <S, M> Replica<S, M> where S: Store, M: StateMachine {
...
+            actions.client_messages.push((from.into(), messages::command_response_unknown_leader()));
         } else if self.is_follower() {
+            scoped_debug!("proposal recieved, but replica is not leader");
```
Comment (danburkert) [inline src/replica.rs]: misspelled received

File: src/replica.rs [hunk 10]
```diff
@@ -481,34 +490,69 @@ impl <S, M> Replica<S, M> where S: Store, M: StateMachine {
...
             let message =
-                messages::proposal_response_not_leader(&self.peers[&self.follower_state.leader.unwrap()]);
+                messages::command_response_not_leader(&self.peers[&self.follower_state.leader.unwrap()]);
             actions.client_messages.push((from, message));
         } else {
+            scoped_debug!("proposal recieved, replica processing");
```
Comment (danburkert) [inline src/replica.rs]: misspelled received

File: src/replica.rs [hunk 11]
```diff
@@ -481,34 +490,69 @@ impl <S, M> Replica<S, M> where S: Store, M: StateMachine {
...
+                        actions.peer_messages.push((peer, message.clone()));
+                    }
                 }
             }
         }
     }
 
+    /// Issue a client query to the Raft replica.
+    fn query_request(&mut self,
+                        from: ClientId,
```
Comment (danburkert) [inline src/replica.rs]: indent

File: src/replica.rs [hunk 12]
```diff
@@ -481,34 +490,69 @@ impl <S, M> Replica<S, M> where S: Store, M: StateMachine {
...
+                        request: query_request::Reader,
+                        actions: &mut Actions) {
+        scoped_debug!("query from Client({})", from);
```
Comment (danburkert) [inline src/replica.rs]: this should probably be trace level

**Comment by danburkert** [comment]
Overall looks really good, just a couple nits. Thanks for tracking down those bugs.

File: src/replica.rs (hunk 10 above)
Comment (Hoverbear) [inline src/replica.rs]: Lol.... Thanks.

**Comment by Hoverbear** [comment]
@danburkert rebased and fixed nits.

**Comment by danburkert** [comment]
@homu r+

**Pull request merged**